src/gs-discovery-feed-content-provider.c

```diff
@@ -159,54 +175,49 @@ search_done_cb (GObject *source,
 		return;
 	}
 
-	app_list_length = gs_app_list_length (list);
-
-	/* sort by kudos, as there is no ratings data by default */
+	/* First, filter out any irrelevant apps. Then randomize the list -
+	 * this gives us a different order for each day as the random seed
+	 * for gs_app_list_randomize changes on a daily basis. Then, sort
+	 * the list, but only use the kudos as the sorting factor. This will
+	 * put "more relevant" apps closer to the top, but still with an
+	 * internally randomized order. */
+	gs_app_list_filter (list, filter_for_discovery_feed_apps, NULL);
+	gs_app_list_randomize (list);
 	gs_app_list_sort (list, search_sort_by_kudo_cb, NULL);
 
+	app_list_length = gs_app_list_length (list);
+
+	/* Now that we have our indices, start building up a variant based
+	 * on those indices */
 	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
-	for (i = 0; i < app_list_length; i++) {
-		g_autofree gchar *app_thumbnail_filename = NULL;
-		GdkPixbuf *pixbuf;
+	for (i = 0; i < MIN (app_list_length, MAX_INSTALLABLE_APPS); ++i) {
+		GsApp *app = GS_APP (gs_app_list_index (list, i));
+		const gchar *app_id = gs_app_get_id (app);
+		g_autofree gchar *app_thumbnail_filename = get_app_thumbnail_cached_filename (app);
+		g_autoptr(GdkPixbuf) pixbuf = gs_app_get_pixbuf (app);
```

Comment:
`gs_app_get_pixbuf()` is `transfer none`: https://github.com/endlessm/gnome-software/blob/master/lib/gs-app.c#L1634
So this code will remove a reference it does not own.

Reply:
Also a good catch! Fixed.